src/cache/local.ml

```diff
@@ -401,79 +397,76 @@ let make ?(root = default_root ())
 
 let duplication_mode cache = cache.duplication_mode
 
-let trimmable stats = stats.Unix.st_nlink = 1
-
-let _garbage_collect default_trim cache =
+let _garbage_collect cache ~trimmed_so_far =
   let root = metadata_store_root cache in
   let metas =
     List.map ~f:(fun p -> (p, Metadata_file.parse p)) (FSSchemeImpl.list ~root)
   in
-  let f default_trim = function
-    | p, Result.Error msg ->
-      cache.warn
-        [ Pp.textf "remove invalid metadata file %s: %s"
-            (Path.to_string_maybe_quoted p)
-            msg
-        ];
-      Path.unlink_no_err p;
-      { default_trim with Trimming_result.trimmed_metafiles = [ p ] }
-    | p, Result.Ok { Metadata_file.files; _ } ->
-      if
-        List.for_all
-          ~f:(fun { File.digest; _ } -> Path.exists (file_path cache digest))
-          files
-      then
-        default_trim
-      else (
-        cache.info
-          [ Pp.textf
-              "remove metadata file %s as some produced files are missing"
-              (Path.to_string_maybe_quoted p)
+  let f trimmed_so_far (path, metadata) =
+    let should_be_removed =
+      match metadata with
+      | Result.Error msg ->
+        cache.warn
+          [ Pp.textf "remove invalid metadata file %s: %s"
+              (Path.to_string_maybe_quoted path)
+              msg
           ];
-        let res =
-          List.fold_left ~init:default_trim
-            ~f:(fun trim f ->
-              let path_in_cache = file_path cache f.File.digest in
-              try
-                let stats = Path.stat path_in_cache in
-                if trimmable stats then (
-                  Path.unlink_no_err path_in_cache;
-                  Trimming_result.add trim ~file:path_in_cache
-                    ~size:stats.st_size
-                ) else
-                  trim
-              with Unix.Unix_error (Unix.ENOENT, _, _) -> trim)
+        true
+      | Result.Ok { Metadata_file.files; _ } ->
+        if
+          List.for_all
+            ~f:(fun { File.digest; _ } -> Path.exists (file_path cache digest))
             files
-        in
-        Path.unlink_no_err p;
-        res
-      )
+        then
+          false
+        else (
+          cache.info
+            [ Pp.textf "remove metadata file %s as it refers to missing files"
+                (Path.to_string_maybe_quoted path)
+            ];
+          true
+        )
+    in
+    if should_be_removed then (
+      let bytes = (Path.stat path).st_size in
+      Path.unlink_no_err path;
+      Trimming_result.add trimmed_so_far ~bytes
+    ) else
+      trimmed_so_far
```

Comment:
Was it not possible to keep the one-pass `fold` approach ? Here we first build a list with the same number of entries as the number of metadata files (which could be big-ish), to then go through it and trim.

Reply:
I believe it's the same one-pass fold. I've pushed a slight refactoring to make it easier to see that `trimmed_so_far` inside the loop shadows the initial `trimmed_so_far` which is passed to the `_garbage_collect` function.

Reply:
Ah, perhaps you meant we can combine the `map` above with the `fold`? Indeed, that's a good idea. Done!